tensorboard/data/server/logdir.rs

```diff
@@ -0,0 +1,384 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Loader for many runs under a directory.
+
+use std::collections::{HashMap, HashSet};
+use std::path::{Path, PathBuf};
+use walkdir::WalkDir;
+
+use crate::commit::Commit;
+use crate::run::RunLoader;
+use crate::types::Run;
+
+/// A loader for a log directory, connecting a filesystem to a [`Commit`] via [`RunLoader`]s.
+pub struct LogdirLoader<'a> {
+    commit: &'a Commit,
+    logdir: PathBuf,
+    runs: HashMap<Run, RunState>,
+}
+
+struct RunState {
+    /// Stateful loader for this run.
+    #[allow(dead_code)] // under construction
+    loader: RunLoader,
+    /// Path to this run's directory relative to the root logdir.
+    relpath: PathBuf,
+    /// Logdir-relative paths to other directories that normalize to the same name as this run
+    /// after lossy Unicode conversion. This is only non-empty when there are multiple filesystem
+    /// paths that are invalid Unicode and collide. For deduplicating warnings.
+    merged_relpaths: HashSet<PathBuf>,
```

Comment:
Subjective: I found it hard to conceptualize this and struggled a bit to understand how to understand this. It did click and ultimately understood what it means. 

Would it be better/worse to call this `collided_relpaths` or something like it?

Reply:
Sure—thanks for the feedback. Renamed per your suggestion, and expanded
docs. Hopefully this helps.